pkg/gadgets/utils/test/helpers.go

```diff
@@ -0,0 +1,159 @@
+	const one = uint32(1)
+
+	mntnsSpec := &ebpf.MapSpec{
+		Name:       "my_filter_map",
+		Type:       ebpf.Hash,
+		KeySize:    8,
+		ValueSize:  4,
+		MaxEntries: 1024,
+	}
+	m, err := ebpf.NewMap(mntnsSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, mountnsid := range mountnsids {
+		if err := m.Put(mountnsid, one); err != nil {
+			m.Close()
+			return nil, err
+		}
+	}
+
+	return m, nil
+}
+
+type Tracer interface {
+	Stop()
+}
+
+type Test[Event any] struct {
+	GetTracer     func(*Info, func(event Event)) Tracer
+	RunnerConfig  *Config
+	GenerateEvent func() error
+	// One of the returned events should match this
+	GetAnyExpectedEvent func(*Info) *Event
+	// Only one event should be returned and should match this
+	GetExpectedEvent func(*Info) *Event
+}
+
+type testRunner[Event any] struct {
+	Fatalf func(format string, args ...any)
+	t      *testing.T
+	tracer Tracer
+	runner *Runner
+}
+
+func NewTestRunner[Event any](t *testing.T) *testRunner[Event] {
+	testRunner := &testRunner[Event]{
+		t: t,
+	}
+
+	testRunner.Fatalf = func(format string, args ...any) {
+		if testRunner.runner != nil {
+			testRunner.runner.Close()
+		}
+
+		if testRunner.tracer != nil {
+			testRunner.tracer.Stop()
+		}
+
+		testRunner.t.Fatalf(format, args...)
+	}
+
+	return testRunner
+}
+
+func (t *testRunner[Event]) Run(tests map[string]Test[Event]) {
+	var err error
+
+	for name, test := range tests {
+		events := []Event{}
+		eventCallback := func(event Event) {
+			events = append(events, event)
+		}
+
+		t.runner, err = NewRunner(test.RunnerConfig)
+		if err != nil {
+			t.Fatalf("%q: failed to create runner: %s", name, err)
+		}
+
+		info := t.runner.Info
+
+		t.tracer = test.GetTracer(info, eventCallback)
+
+		if err := t.runner.Run(test.GenerateEvent); err != nil {
+			t.Fatalf("error generating event: %s", err)
+		}
+
+		// Give some time for the tracer to capture the events
+		time.Sleep(100 * time.Millisecond)
+
+		t.runner.Close()
+		t.runner = nil
+		t.tracer.Stop()
+		t.tracer = nil
+
+		if test.GetAnyExpectedEvent != nil {
+			found := false
+			expectedEvent := test.GetAnyExpectedEvent(info)
+
+			for _, event := range events {
+				if reflect.DeepEqual(expectedEvent, &event) {
+					found = true
+					break
+				}
+			}
+
+			if !found {
+				// Provide extra info when only a single event was captured
+				if len(events) == 1 {
+					t.Fatalf("%q: event doesn't match:\n%s", name,
+						cmp.Diff(expectedEvent, &events[0]))
+				}
+				t.Fatalf("%q: event wasn't captured", name)
+			}
+		} else if test.GetExpectedEvent != nil {
+			expectedEvent := test.GetExpectedEvent(info)
+			if len(events) != 1 {
+				t.Fatalf("%q: more than one event found", name)
+			}
+
+			if !reflect.DeepEqual(expectedEvent, &events[0]) {
+				t.Fatalf("%q: event doesn't match:\n%s", name,
+					cmp.Diff(expectedEvent, &events[0]))
+			}
+		} else {
+			if len(events) != 0 {
+				t.Fatalf("%q: no events are expected", name)
+			}
+		}
```

Comment:
Perhaps this huge if/else if/else block could be broken down in 3 or more smaller functions to favor use of function return etc for flow control rather than else? This way code becomes not so deeply intended and easier to understand.